python/cudf/cudf/core/indexed_frame.py

```diff
@@ -1694,6 +1694,204 @@ def last(self, offset):
+            if frac > 1 and not replace:
+                raise ValueError(
+                    "Replace has to be set to `True` when upsampling the "
+                    "population `frac` > 1."
+                )
+            if n is not None:
+                raise ValueError(
+                    "Please enter a value for `frac` OR `n`, not both."
+                )
+            n = int(round(size * frac))
+
+        if n > 0 and size == 0:
+            raise ValueError(
+                "Cannot take a sample larger than 0 when axis is empty."
+            )
+
+        if isinstance(random_state, cp.random.RandomState):
+            lib = cp
+        elif isinstance(random_state, np.random.RandomState):
+            lib = np
+        else:
+            # Construct random state if `random_state` parameter is None or a
+            # seed. By default, cupy random state is used to sample from rows
+            # and numpy is used to sample from columns. In general, cuDF
+            # assumes the number of columns is much smaller than the number of
+            # rows, thus using numpy random states can avoid kernel launching
+            # overhead on creating a small gather map.
+            lib = cp if axis == 0 else np
+            random_state = lib.random.RandomState(seed=random_state)
+
+        # Normalize `weights` array.
+        if weights is not None:
+            if isinstance(weights, str):
+                raise NotImplementedError(
+                    "Weights specified by string is unsupported yet."
+                )
+
+            if size != len(weights):
+                raise ValueError(
+                    "Weights and axis to be sampled must be of same length."
+                )
+
+            weights = lib.asarray(weights)
+            weights = weights / weights.sum()
+
+        if axis == 0:
+            return self._sample_axis_0(
+                n, weights, replace, random_state, ignore_index
+            )
+        else:
+            if isinstance(random_state, cp.random.RandomState):
+                raise ValueError(
+                    "Sampling from `axis=1`/`columns` with cupy random state"
+                    "isn't supported."
+                )
+            return self._sample_axis_1(
+                n, weights, replace, random_state, ignore_index
+            )
+
+    def _sample_axis_0(
+        self,
+        n: int,
+        weights: Optional[ColumnLike],
+        replace: bool,
+        random_state: Union[np.random.RandomState, cp.random.RandomState],
+        ignore_index: bool,
+    ):
+        try:
+            gather_map_array = random_state.choice(
+                len(self), size=n, replace=replace, p=weights
+            )
+        except NotImplementedError as e:
+            raise NotImplementedError(
+                "Unsupported arguments for gather map generation."
+            ) from e
+
+        return self._gather(
+            cudf.core.column.as_column(gather_map_array),
```

Comment:
Just wondering, is there something faster we can do here than calling `as_column` since we know the input is a numpy/cupy array? Maybe this is the easiest solution since it will automatically dispatch for `__array_interface__` vs `__cuda_array_interface__`, just thought I'd check. No worries if this is a no-op.

Reply:
This is what I wondered as well, if we have a faster code path for numpy/cupy array. Let me do some benchmark first. We can make a `build_column_from_array` factory function to capture this idea if it works.

Reply:
I wouldn't go too far with this for now, I think it's out of scope for this PR if nothing exists already. At some point we'll need to refactor `as_column`, and that will open up a lot of these kinds of things.

Reply:
Unfortunately from `as_column` it seems the conversion from cupy->cudf column isn't that straightforward: https://github.com/rapidsai/cudf/blob/a72479f2ab7d2ef7b9c8c10df7835434894054dc/python/cudf/cudf/core/column/column.py#L1760-L1807, will put this as no-op for now.